From 7bcd07e636f57e79b8338f73e422e30b514b98b1 Mon Sep 17 00:00:00 2001
From: Jakub Pawlowski <jpawlowski@google.com>
Date: Mon, 16 Nov 2020 15:54:00 +0100
Subject: [PATCH 3/3] [Q_asb_2021-02] Check permission before sending batch
 scan result

Use same checks as for regular scan results

Bug: 172670415
Test: compilation
Merged-In: I4274026943ce64a51a30c3fbf6cc85eec853ad4f
Change-Id: I4274026943ce64a51a30c3fbf6cc85eec853ad4f
(cherry picked from commit f47a016eecb7e0e534b77219182607f4a7fd832c)
---
 .../android/bluetooth/gatt/GattService.java   | 26 +++++++++++++++++++
 1 file changed, 26 insertions(+)

diff --git a/src/com/android/bluetooth/gatt/GattService.java b/src/com/android/bluetooth/gatt/GattService.java
index bddb0a50d..47c14ec66 100644
--- a/src/com/android/bluetooth/gatt/GattService.java
+++ b/src/com/android/bluetooth/gatt/GattService.java
@@ -1563,6 +1563,15 @@ public class GattService extends ProfileService {
         mScanManager.callbackDone(clientIf, status);
     }
 
+    ScanClient findBatchScanClientById(int scannerId) {
+        for (ScanClient client : mScanManager.getBatchScanQueue()) {
+            if (client.scannerId == scannerId) {
+                return client;
+            }
+        }
+        return null;
+    }
+
     void onBatchScanReports(int status, int scannerId, int reportType, int numRecords,
             byte[] recordData) throws RemoteException {
         if (DBG) {
@@ -1577,6 +1586,17 @@ public class GattService extends ProfileService {
             if (app == null) {
                 return;
             }
+
+            ScanClient client = findBatchScanClientById(scannerId);
+            if (client == null) {
+                return;
+            }
+
+            // Do not report if location mode is OFF or the client has no location permission
+            if (!hasScanResultPermission(client)) {
+                return;
+            }
+
             if (app.callback != null) {
                 app.callback.onBatchScanResults(new ArrayList<ScanResult>(results));
             } else {
@@ -1619,6 +1639,12 @@ public class GattService extends ProfileService {
             Log.e(TAG, "app not found from id(" + client.scannerId + ") for received callback");
             return;
         }
+
+        // Do not report if location mode is OFF or the client has no location permission
+        if (!hasScanResultPermission(client)) {
+            return;
+        }
+
         if (client.filters == null || client.filters.isEmpty()) {
             sendBatchScanResults(app, client, new ArrayList<ScanResult>(allResults));
             // TODO: Question to reviewer: Shouldn't there be a return here?
-- 
2.34.1

